Don't mix up glyph IDs when multiple installed fonts share a name

fantasai opened the pull request:
#### d621347fbb522a61863862482451ae76e47d696f
<pre>
Don&apos;t mix up glyph IDs when multiple installed fonts share a name
<a href="https://bugs.webkit.org/show_bug.cgi?id=167068">https://bugs.webkit.org/show_bug.cgi?id=167068</a>
<a href="https://rdar.apple.com/129891005">rdar://129891005</a>

Reviewed by Vitor Roriz.

On CoreText platforms, we store two CTFontRefs for a given Font.
If they end up pointing to different fonts, this causes major problems
such as mixing up glyph IDs and rendering nonsense text.

This patch merges them into a single object by
* Dropping the LastResort CoreText fallback, which seems no longer needed.
* Applying the width variants for all uses, which should be more correct anyway.
* Updating all callers accordingly.

* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::getCTFont const):
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::ctFont const):
(WebCore::FontPlatformData::font const): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::systemFallbackForCharacterCluster):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::primaryFontIsSystemFont const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::applyTransforms const):
(WebCore::Font::determinePitch):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformSupportsCodePoint const):
(WebCore::Font::isProbablyOnlyUsedToRenderIcons const):
(WebCore::Font::otSVGTable const):
(WebCore::Font::hasComplexColorFormatTables const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::mapFontWidthVariantToCTFeatureSelector):
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::toIPCData const):
(WebCore::cascadeToLastResortAttributesDictionary): Deleted.
(WebCore::cascadeToLastResortAndVariationsFontDescriptor): Deleted.
(WebCore::FontPlatformData::ctFont const): Deleted.
* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):

Canonical link: <a href="https://commits.webkit.org/281022@main">https://commits.webkit.org/281022@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ec8228b27017e6dd988d89dc91559d10f3f9c87

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37751 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10908 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62049 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60553 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45387 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9065 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47322 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6330 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60455 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35361 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50516 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28166 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/32116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7815 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7872 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/54069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63753 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2337 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/8118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54641 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50538 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54707 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1979 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8708 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/33580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34666 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34411 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz reviewed (commented):
Can you add to the commit message the differences between the 2 CTFont's that we store currently?

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -46,6 +66,19 @@ FontPlatformData::FontPlatformData(RetainPtr<CTFontRef>&& font, float size, bool
+        RetainPtr<CTFontRef> newFont = adoptCF(CTFontCreateWithFontDescriptor(newDescriptor.get(), m_size, 0));
+
+        if (newFont)
+            m_font = newFont;
```
I wonder if there is a way to do this before we instantiate m_font for the first time (when we create FontPlatformData) so we avoid doing an extra `CTFontCreateWithFontDescriptor`

fantasai replied:
I thought about that, and I agree that long-term that's what we should do, but it would complicate this patch a lot and I didn't want to make those changes. This works, and it's not a regression from what we've been doing.

vitorroriz replied:
Sounds reasonable but then let's create a radar with a FIXME here so we don't forget about it.

fantasai replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=276635

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -119,72 +152,6 @@ CTFontRef FontPlatformData::registeredFont() const
-
-static RetainPtr<CFDictionaryRef> cascadeToLastResortAttributesDictionary()
-{
-    auto lastResort = adoptCF(CTFontDescriptorCreateWithNameAndSize(CFSTR("LastResort"), 0));
```
Is it right to no longer create the font with the "LastResort" postscript name?

If so, shouldn't we remove any mentions to it on code?

For example, we have special handling for it here: https://searchfox.org/wubkat/source/Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm#254

vitorroriz replied:
Actually we are not creating a font with such a name. What we do is:

1. `CTFontDescriptorCreateWithNameAndSize` creates a font descriptor with `LastResort` postscript name
2. creates a dictionary that maps from `kCTFontCascadeListAttribute` attribute to an array with a single descriptor, the `LastResort` descriptor in [1]
3. `CTFontDescriptorCreateWithAttributes` uses the dictionary in [2] to create a font descriptor that has `kCTFontCascadeListAttribute` attribute set to a list with a single descriptor, `LastResort` descriptor.
4. `FontPlatformData::ctFont()` will create a new `CTFont` that is a copy of the existing `CTFont` held at `m_font` but now with its `kCTFontCascadeListAttribute` pointing to a list with a single font descriptor, the `LastResort` descriptor
5. if the requested `FontWidthVariant` is not `::RegularWidth` we will create a new font descriptor based on the `CTFont` at [4], and then a new `CTFont` based on such descriptor. We will hold such font at `m_ctFont` and this is used at some points in our code base, until this patch. I don't really understand why we need this, but it seems that after this patch, the only difference is that the `CTFont` we return will no longer have its `kCTFontCascadeListAttribute` attribute pointing to a list containing the `LastResort` font descriptor.

fantasai replied:
> If so, shouldn't we remove any mentions to it on code?

No, if we run into LastResort it means that CoreText couldn't find a font that can render the glyph, and we do want to continue to have special handling for that case.

vitorroriz replied:
Ok, so the question we have to answer here is why did we need a `CTFont` that  has its `kCTFontCascadeListAttribute` set to a list containing a single descriptor (the one from `LastResort`).

fantasai replied:
The `FontWidthVariant` stuff is used for implementing `text-combine-horizontal` in https://searchfox.org/wubkat/source/Source/WebCore/rendering/RenderCombineText.cpp#139
AFAICT applying it to both m_font and m_ctFont would have been the right thing to do; we're just lucky that the various places where we looked up m_font instead of m_ctFont would (in any given font, assuming they're the same font like they're supposed to be) return the same values. I checked the relevant `text-combine-horizontal` effects, and they work as expected.

fantasai replied:
> No, if we run into LastResort it means that CoreText couldn't find a font that can render the glyph, and we do want to continue to have special handling for that case.

(Although to be honest, I have no idea how to run into this case. We don't seem to be hitting it when we render unknown codepoints.)

fantasai replied:
> Ok, so the question we have to answer here is why did we need a CTFont that has its kCTFontCascadeListAttribute set to a list containing a single descriptor (the one from LastResort).

According to Myles, it was because we used to check for LastResort in order to know whether a font needed fallback in complex text. But our fallback behavior seems to work just fine through multiple fonts with the patch enabled. (We do have a bug when we hit a generic keyword, but that's a pre-existing bug.)

vitorroriz replied:
I guess to make sure we would need a test in which we use the `CodePath::Complex` path (we can use specific codepoints, for example) and use a `font-family` list in which the first font does not support a code point but the second does, and then we use as `-ref` the text but with a `font-family` list composed only by the second font in the test's `font-family` list. We can use Ahem as the second font to make sure it does not coincide with a font that CT would fallback to.

fantasai replied:
Added fast/text/font-fallback.html

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
```diff
@@ -316,7 +282,7 @@ FontPlatformData::FontPlatformData(float size, WebCore::FontOrientation&& orient
 FontPlatformData::IPCData FontPlatformData::toIPCData() const
 {
-    auto ctFont = font();
+    auto ctFont = FontPlatformData::ctFont();
```
why not simply `ctFont()`?

fantasai replied:
It wouldn't compile. I can also rename `ctFont` to `font` and it will compile.

vitorroriz replied:
ah, right, you can't use this-> because it is a static member, don't worry about it.

fantasai replied:
I mean, I don't mind. :) I think it would be a minor improvement in readability, it just would tweak more of the lines in the diff. :p

vitorroriz commented on LayoutTests/fast/text/font-fallback-expected.html:
```diff
@@ -0,0 +1,16 @@
+
+</style>
+
+<p><span>A</span>䁠<span>A</span>Ψ
```
isn't `七` missing here?

fantasai replied:
I'm using A of Ahem instead.

fantasai committed:
Don't mix up glyph IDs when multiple installed fonts share a name
https://bugs.webkit.org/show_bug.cgi?id=167068
rdar://129891005

Reviewed by Vitor Roriz.

On CoreText platforms, we store two CTFontRefs for a given Font.
If they end up pointing to different fonts, this causes major problems
such as mixing up glyph IDs and rendering nonsense text.

This patch merges them into a single object by
* Dropping the LastResort CoreText fallback, which seems no longer needed.
* Applying the width variants for all uses, which should be more correct anyway.
* Updating all callers accordingly.

* Source/WebCore/platform/graphics/Font.h:
(WebCore::Font::getCTFont const):
* Source/WebCore/platform/graphics/FontPlatformData.h:
(WebCore::FontPlatformData::ctFont const):
(WebCore::FontPlatformData::font const): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::systemFallbackForCharacterCluster):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::primaryFontIsSystemFont const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::platformInit):
(WebCore::Font::platformCharWidthInit):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):
(WebCore::Font::platformWidthForGlyph const):
(WebCore::Font::applyTransforms const):
(WebCore::Font::determinePitch):
(WebCore::Font::platformBoundsForGlyph const):
(WebCore::Font::platformPathForGlyph const):
(WebCore::Font::platformSupportsCodePoint const):
(WebCore::Font::isProbablyOnlyUsedToRenderIcons const):
(WebCore::Font::otSVGTable const):
(WebCore::Font::hasComplexColorFormatTables const):
* Source/WebCore/platform/graphics/coretext/FontPlatformDataCoreText.cpp:
(WebCore::mapFontWidthVariantToCTFeatureSelector):
(WebCore::FontPlatformData::FontPlatformData):
(WebCore::FontPlatformData::updateSize):
(WebCore::FontPlatformData::toIPCData const):
(WebCore::cascadeToLastResortAttributesDictionary): Deleted.
(WebCore::cascadeToLastResortAndVariationsFontDescriptor): Deleted.
(WebCore::FontPlatformData::ctFont const): Deleted.
* Source/WebCore/platform/graphics/mac/ComplexTextControllerCoreText.mm:
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):

Canonical link: https://commits.webkit.org/281022@main
